### DISCUSSION TITLE
generate secrets from other secrets?

### DISCUSSION THREAD

**Discussion by airtonix** [fluxcd/flux2#3167] [Q&A]
I have some secrets in another repo that get pulled in.

Currently they end up in the `flux-system` namespace.

For now, I'd like to keep them there.

But my `cert-manager` namespace needs some secrets from them.

I'd like to use secretGenerator to pick some keys out of these and generate a `cert-manager` scoped secret.

However, I'm not sure that's possible so currently I think my only alternative is to start creating more segregation in my secrets repo and then create more sync kustomizations that pull from those separate directories and put them into `targetNamespace: cert-manager`.

I don't like this since it means there are some secrets that many systems need access to, but for whatever reason they only look for secrets you specify withing their local namespaces.

**Comment by kingdonb** [comment]
Includes might solve this – you can mount a gitrepository inside another gitrepository, and that will help where you need to bridge secrets into a separate repo that doesn't have any secrets but depends on them to exist in a certain location (perhaps you maintain your local copy via .gitignore unencrypted for use in the dev environment setting)

The fundamental problem is that you can't reach across one git repo into another with a configmap generator, or secretgenerator, in Kustomize controller. But I think you could do it once you have used `GitRepositories.spec.include` to locate those decrypted secrets in the same payload, where the Kustomize controller can handle 

It's normal that secrets are only available in the same namespace, that's how Kubernetes multi-tenancy is architected from the start. Secrets live in a namespace, only cluster admins can reach outside of their own namespace(s) as they are permitted via RBAC.

It seems like it would be super inconvenient to manage secrets this way, unless it just happens to fit your process. There is a policy tool for everything these days, and external-secrets operator does this (copy secrets across namespaces) so also Kyverno does this.

You will generally need for the secret and the rest of the manifest source to be located together one way or another in order to use them with a configmap generator, (IOW you can't use configmap generator with external secrets, unless there's some trick I haven't thought of, that's why SOPS is the recommended solution, because it works with generators and it can be secured.)

You can of course always do this without a kustomize generator as an alternative, the Helm controller will notice if a secretRef has changed its content even if the secret wasn't bundled with the helmrelease, it does this on the HelmRelease reconcile interval. Just to point out that you don't necessarily need to use the configmap generator feature to get it done.

Finally while we were briefly discussing this at Flux bug scrub last week, (sorry I haven't got back to you sooner) we found this while we were discussing: https://cert-manager.io/docs/tutorials/syncing-secrets-across-namespaces/

Since you mentioned cert-manager, we found that it actually has a guide for syncing secrets across namespaces! This might be exactly what you needed. Surely one of these solutions can help. Thanks for posting the question and thanks for using Flux!
